### ISSUE TITLE
const_err does not trigger on associated constants

### ISSUE THREAD

**Issue by Nemo157** [rust-lang/rust#71282]
I tried this code:

```rust
pub struct Foo;

pub const FOO: Foo = Foo::new(1);

impl Foo {
    pub const FOO: Self = Self::new(1);

    pub const fn new(i: usize) -> Foo {
        [][i] = 5;
        Foo
    }
}
```

I expected to see this happen: Both `FOO` and `Foo::FOO` should trigger `const_err` errors.

Instead, this happened: Only `FOO` was marked with `const_err`

### Meta

Checked on current stable (1.42) and nightly (2020-04-17) on the playground

**Comment by Nemo157** [comment]
This was discussed at https://github.com/rust-lang/rust/issues/67191#issuecomment-570433895 and below, so might be covered by https://github.com/rust-lang/rust/pull/70820 but I didn't really look into the details.

**Comment by RalfJung** [comment]
Cc @rust-lang/wg-const-eval 

I would say this is expected. The error comes from actually computing the const (not some kind of static analysis). Associated consts cannot in general be evaluated eagerly as they could depend on generic parameters, unlike global consts.

You will [still get the error if you actually *use* `Foo::FOO`](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=00bfa3d782f4123d9b1680178601daf7).

I suppose we could try to still eagerly evaluate monomorphic associated consts like this one, but that scheme will also quickly reach its limit, and it feels rather ad-hoc. Currently, from what I recall, const-prop and const-eval carefully balance who is analyzing what to avoid duplicate warnings; any such change would need careful testing to make sure we do not regress that.
